packages/live-share-inking/src/canvas/DryWetCanvas.ts

```diff
@@ -0,0 +1,269 @@
+        }
+
+        if (this.rendersProgressively()) {
+            this._pendingPointsStartIndex = this._points.length;
+        }
+    }
+
+    protected internalBeginStroke(p: IPointerPoint) {
+        this._points = [p];
+        this._pendingPointsStartIndex = 0;
+    }
+
+    protected internalAddPoint(p: IPointerPoint) {
+        this._points.push(p);
+    }
+
+    protected internalEndStroke(p?: IPointerPoint) {
+        if (p) {
+            this._points.push(p);
+        }
+    }
+
+    protected internalCancelStroke() {
+        // Nothing to do in base implementation
+    }
+
+    protected get points(): IPointerPoint[] {
+        return this._points;
+    }
+}
+
+/**
+ * Represents a canvas suitable for "dry ink", i.e. the persistent drawing. DryCanvas renders
+ * synchonously.
+ */
+export class DryCanvas extends DryWetCanvas {
+    protected rendersAsynchronously(): boolean {
+        return false;
+    }
+
+    /**
+     * A "dry" canvas renders multiple strokes which might each have a different brush.
+     * The underlying context's global alpha and composite operation must be set before
+     * rendering each stroke.
+     * @param context The context to set the blend mode on, given the current brus.
+     */
+    protected adjustOpacity(context: CanvasRenderingContext2D) {
+        switch (this.brush.type) {
+            case "laser":
+                context.globalAlpha = InkingCanvas.laserShadowOpacity;
+                break;
+            case "highlighter":
+            default:
+                super.adjustOpacity(context);
+                break;
+        }
+    }
+}
+
+/**
+ * Represents a canvas suitable for "wet ink", i.e. an ongoing stroke.
+ */
+export class WetCanvas extends DryWetCanvas {
+    /**
+     * This setting exists for testing purposes only. Do not set to `false`
+     * in production code as it will drastically impair performance.
+     * Asynchronous rendering, combined with progressive rendering,
+     * causes two canvases rendering the same stroke (with the exact same
+     * points) to not render the exact same pixels.
+     */
+    private static readonly forceSynchronousRendering = false;
+
+    protected rendersAsynchronously(): boolean {
+        return WetCanvas.forceSynchronousRendering ? false : super.rendersAsynchronously();
+    }
+
+    protected rendersProgressively(): boolean {
+        return WetCanvas.forceSynchronousRendering ? false : super.rendersProgressively();
+    }
+
+    protected getBrushCssColor(): string {
+        // In a wet canvas, when using the highlighter, the brush color
+        // used to draw the stroke is always opaque, and it's the canvas
+        // itself that is semi-transparent.
+        return toCssColor(this.brush.color);
+    }
+
+    /**
+     * A "wet" canvas always renders a single stroke and is discarded when that stroke end.
+     * It needs to be properly composited on whatever other DOM it is overlyed on, basically
```

Comment:
```suggestion
     * It needs to be properly composited on whatever other DOM it is overlayed on, basically
```

Reply:
I'm starting to feel embarrassed now :)

Reply:
haha sorry, it's one of the things that I nitpick...